### ISSUE TITLE
Pb in result

### ISSUE THREAD

**Issue by cyril-data** [cyril-data/sf_continuous_action#2]
# algo :  
- actor / critic / sf critic : s**ame feature** (`share_features_extractor = True`)

- **actor 's latent has  net_arch layer.** critic and sf don't.  https://github.com/cyril-data/sf_continuous_action/blob/243bb2f69338dbc5ab908753d34fc23e8ca04f3a/ppo_sf/policies.py#L270C1-L274C60 . 
```
        if isinstance(net_arch, dict):
            net_arch["vf"] = []
            net_arch["sf"] = []
        else:
            net_arch = {"pi": net_arch, "vf": [], "sf": []}
```
# XP

- training SF learning with random policy 
- freeze feature and train classic ppo with sf feature. 
- Critic Value use frozen` psi = self.sf_net` ,  to compute value fonction https://github.com/cyril-data/sf_continuous_action/blob/1aa7be08aa21d6c7e2e86086c0dd32c380f9476d/ppo_sf/policies.py#L447C1-L450C37 : 
```
        psi = self.sf_net(latent_sf)
        if self.freeze_feature:
            psi = psi.detach()
        values = self.value_net(psi)
```
- compare `ep_rew_mean` with a classic ppo loss and with feature no frozen  (

=> pb: classic ppo is slighlty better than frozen SF feature. 
=> more xp need to be done

**Comment by cyril-data** [comment]
exemple of SF freeze (gray) and classic no freeze ppo (green) :  

<img width="248" height="499" alt="Image" src="https://github.com/user-attachments/assets/fdf30177-0938-4067-b30f-6d8df215d573" />

**Comment by cyril-data** [comment]
**archi : nl2_fd64**
```
        if isinstance(net_arch, dict):
            net_arch["vf"] = []
            net_arch["sf"] = []
        else:
            net_arch = {"pi": [], "vf": [], "sf": []}
```

value detach : 
```
    def obs_to_value_sf_latent_pi(self, obs):
        latent_pi, latent_vf, latent_sf = self.extract_latent(obs)

        psi = self.sf_net(latent_sf)
        if self.freeze_feature:
            psi = psi.detach()
        values = self.value_net(psi)
        return values, psi, latent_pi
```

<img width="306" height="420" alt="Image" src="https://github.com/user-attachments/assets/fad3ae99-8cf0-4dd5-9e2e-f749132621b0" />

**Comment by cyril-data** [comment]
**archi : nl3_fd64**

```
        if isinstance(net_arch, dict):
            net_arch["vf"] = []
            net_arch["sf"] = []
        else:
            net_arch = {"pi": [], "vf": [], "sf": []}
```

value detach :

```
    def obs_to_value_sf_latent_pi(self, obs):
        latent_pi, latent_vf, latent_sf = self.extract_latent(obs)

        psi = self.sf_net(latent_sf)
        if self.freeze_feature:
            psi = psi.detach()
        values = self.value_net(psi)
        return values, psi, latent_pi
```


<img width="256" height="384" alt="Image" src="https://github.com/user-attachments/assets/328b4a9f-9fd7-40d3-818c-5972b6ff5868" />

**Comment by cyril-data** [comment]
archi : 
- **nl3_fd64 for feature**
- **pi_v_6464** for value and pi but value is detach and compute fron psi :

```
    def obs_to_value_sf_latent_pi(self, obs):
        latent_pi, latent_vf, latent_sf = self.extract_latent(obs)

        psi = self.sf_net(latent_sf)
        if self.freeze_feature:
            psi = psi.detach()
        values = self.value_net(psi)
        return values, psi, latent_pi
```

<img width="306" height="464" alt="Image" src="https://github.com/user-attachments/assets/e6b61554-f783-4ce8-9bd3-9133d7603a12" />

**Comment by cyril-data** [comment]
archi :   
-  **nl3_fd64 for feature**
-  **pi_v_6464** for value and pi but value **is not detach** and **compute from its feature** :

   ```
    def obs_to_value_sf_latent_pi(self, obs):
        latent_pi, latent_vf, latent_sf = self.extract_latent(obs)

        psi = self.sf_net(latent_sf)
        # if self.freeze_feature:
        #     psi = psi.detach()
        values = self.value_net(latent_vf)
        return values, psi, latent_pi
```

**Comment by cyril-data** [comment]
Result are good with value computed from sf feature, but with 64,64 latent layer above  :

<img width="306" height="464" alt="Image" src="https://github.com/user-attachments/assets/ac346cc7-3b21-42c6-abff-d6a20198ee22" />

**Comment by cyril-data** [comment]
convergence in value loss seems more stable : 

<img width="333" height="439" alt="Image" src="https://github.com/user-attachments/assets/bd7d3cf9-939b-4f1a-a519-c22c7a8a8e6f" />

**Comment by cyril-data** [comment]
It looks like Value function calculated directly (linearly) from **Pure exploratory Linear Successor Feature** (psi is linear from feature phi, and SF is trained with pure random policy) do not achieve to converge properly, and the actor policy do not get to the optimal policy. 

=> But from **Pure exploratory Linear Successor Feature**, is it possible though, to get good enough value convergence to reach optimal policy if multiple perception neural layers are added from the feature phi obtained by SF (frozen when value training).

**Comment by cyril-data** [comment]
- {'pi': [64, 64], 'vf': [64], 'sf': []}" --fn "nn.Identity" => **do not** reach optimal policy but more slowly 
- {'pi': [64, 64], 'vf': [64], 'sf': []}" --fn "nn.Tanh" => **reach optimal policy** but _more slowly_
- {'pi': [64,64], 'vf': [64,64], 'sf': []}" --fn "nn.Identity" => **reach optimal policy**

**Comment by cyril-data** [comment]
**--nl 3 --fd 64 --it 70000 --na "{'pi': [64,64], 'vf': [64], 'sf': []}" --fn "nn.Identity"** 

**Without r_loss** is not so bad ! 

<img width="274" height="482" alt="Image" src="https://github.com/user-attachments/assets/184e6ede-c7f6-4b18-99cd-668ddc95f74f" />
